### PR TITLE
Kosarev Egor. Lab 4. Option 2

### PR DESCRIPTION
**Pull request by EgorKosarevv**

### CODE REVIEW

**Review by aobolensk — APPROVED** [review]

File: mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+namespace {
+class KosarevEgorFMAPass
+    : public PassWrapper<KosarevEgorFMAPass, OperationPass<ModuleOp>> {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]: Please narrow down operation that current pass is scheduled on.

File: mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      mlir::Value fmaResult =
+          builder.create<mlir::math::FmaOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), thirdOperand);
+      addOp.replaceAllUsesWith(fmaResult);
+      addOp.erase();
+      mulOp.erase();
+    };
+
+    module.walk([&](mlir::Operation *op) {
+      if (auto addOp = llvm::dyn_cast<mlir::LLVM::FAddOp>(op)) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]:
```suggestion
    module.walk([&](mlir::LLVM::FAddOp faddOp) {
```

File: mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        mlir::Value addLhs = addOp.getOperand(0);
+        mlir::Value addRhs = addOp.getOperand(1);
+
+        if (!addLhs.getType().isa<mlir::FloatType>() ||
+            !addRhs.getType().isa<mlir::FloatType>()) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]: This is not necessary if your operation is FAdd already.

File: mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          for (auto &use : value.getUses()) {
+            if (use.getOwner() != userOp) {
+              return false;
+            }
+          }
+          return true;
+        };
+
+        if (auto mulOp = addLhs.getDefiningOp<mlir::LLVM::FMulOp>()) {
+          if (isSingleUse(mulOp->getResult(0), addOp)) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]:
```suggestion
          if (mulOp->hasOneUse()) {
```

Comment (m-ly4) [reply mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]: The same is for second branch.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp (hunk 1 above)
Comment (m-ly4) [reply mlir/lib/Transforms/lab4/kosarev_egor/KosarevEgorFMAPass.cpp]: Don't forget to apply the remark above also.

**Review by aobolensk — APPROVED** [review]

**Review by m-ly4 — APPROVED** [review]

**Comment by m-ly4** [comment]
@EgorKosarevv Could you please re-create MR and fix your tests?

**Pull request merged**